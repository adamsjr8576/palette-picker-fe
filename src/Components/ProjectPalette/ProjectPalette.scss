.project-colors-article {
  display: flex;
  justify-content: space-around;
  align-items: center;
  margin: 10px 0px;
  width: 80%;
}

.colors-div-container {
  display: flex;
  flex-direction: column;
  align-items: center;
  width: 75%;
}

.color-box {
  height: 20px;
  width: 100%;
}