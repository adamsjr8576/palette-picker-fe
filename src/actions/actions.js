export const addProject = (project) => ({
  type: 'ADD_PROJECT',
  project
})