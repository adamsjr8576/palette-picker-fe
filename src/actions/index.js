export const addNewPalette = currentPalette => ({
  type: 'ADD_PALETTE',
  currentPalette
});

export const updatePaletteLocked = hexCode => ({
  type: 'UPDATE_PALETTE_LOCKED',
  hexCode
});

export const addInitialProjects = projects => ({
  type: 'ADD_INITIAL_PROJECTS',
  projects
});

export const addPalettes = palettes => ({
  type: 'ADD_PALETTE_TO_PROJECT',
  palettes
});

export const addProject = project => ({
  type: 'ADD_PROJECT',
  project
});

export const deletePaletteFromStore = palette => ({
  type: 'DELETE_PALETTE',
  palette
});

export const deleteProjectFromStore = project => ({
  type: 'DELETE_PROJECT',
  project
});
