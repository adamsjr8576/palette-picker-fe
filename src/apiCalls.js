export const postProject = (projectName) => {
  const options = {
    method: 'POST',
    body: JSON.stringify({
      name: projectName
    }),
    headers: {
      'Content-Type': 'application/json'
    }
  }

  return fetch(process.env.REACT_APP_BACKEND_URL + '/api/v1/projects', options)
    .then(res => res.json());
}

export const getProjectById = (id) => {
  return fetch(process.env.REACT_APP_BACKEND_URL + `/api/v1/projects/${id.id}`)
    .then(res => res.json());
}

export const getProjects = () => {
  return fetch(process.env.REACT_APP_BACKEND_URL + `/api/v1/projects`)
    .then(res => res.json());
}

export const postPalette = (body) => {
  const options = {
    method: 'POST',
    body: JSON.stringify(body),
    headers: {
      'Content-Type': 'application/json'
    }
  }

  return fetch(process.env.REACT_APP_BACKEND_URL + '/api/v1/palettes', options)
    .then(res => res.json());
}

export const getPaletteById = (id) => {
  return fetch(process.env.REACT_APP_BACKEND_URL + `/api/v1/palettes/${id.id}`)
    .then(res => res.json());
}

export const deletePalette = (id) => {
  return fetch(process.env.REACT_APP_BACKEND_URL + `/api/v1/palettes/${id}`)
    .then(res => res.json())
}