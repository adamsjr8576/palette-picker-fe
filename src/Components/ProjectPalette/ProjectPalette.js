import React from 'react';
import './ProjectPalette.scss';
import images from '../../images/images';
import { connect } from 'react-redux';
import { deletePalette } from '../../apiCalls';
import { deletePaletteFromStore } from '../../actions/index';

export const ProjectPalette = ({ name, id, colors, deletePaletteFromStore, palettes }) => {

  const getColors = colors.map(color => {
    return(
      <div className='color-box' style={{ backgroundColor: `${color}` }} key={color}>
        <p>{color}</p>
      </div>
    )
  })

  const removePalette = (id) => {
    const paletteToDelete = palettes.find(item => item.id === id)
    deletePalette(id, paletteToDelete)
    .then(res => console.log(res));

    deletePaletteFromStore(paletteToDelete)
  }

  return(
    <article className='project-colors-article'>
      <div>
        <h3>{name}</h3>
        <button data-testid='delete-button' className='trash-btn' onClick={ () => removePalette(id) }><img className='icon-bin' src={images.bin} /></button>
      </div>
      <div className='colors-div-container'>
        {getColors}
      </div>
    </article>
  )
}

export const mapStateToProps = state => ({
  palettes: state.palettes
});

export const mapDispatchToProps = dispatch => ({
  deletePaletteFromStore: palette => dispatch( deletePaletteFromStore(palette) )
});

export default connect(mapStateToProps, mapDispatchToProps)(ProjectPalette);
