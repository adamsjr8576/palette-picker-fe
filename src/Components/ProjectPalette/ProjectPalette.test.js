import React from 'react';
import { ProjectPalette, mapDispatchToProps, mapStateToProps } from './ProjectPalette';
import { deletePaletteFromStore } from '../../actions/index';
import { render, fireEvent } from '@testing-library/react';
import { deletePalette } from '../../apiCalls';

jest.mock('../../apiCalls.js');

describe('ProjectPalette', () => {
  let mockProps, setup;

  describe('ProjectPalette Component', () => {

    beforeEach(() => {
      deletePalette.mockImplementation(() => {
        return Promise.resolve('Successful delete!');
      });
    });

    test('Should match the snapshot', () => {
      mockProps = {
        name: 'Party Palette',
        id: 15,
        colors: ['#11232', '#11222', '#34232', '#11562', '#11852'],
        deletePaletteFromStore: jest.fn(),
        palettes: [{
          colors: ["#c005d4", "#57a3d0", "#88b8b1", "#7bac74", "#e6488b"],
          id: 21,
          name: "Candy",
          created_at: "2020-02-11T21:07:35.955Z",
          updated_at: "2020-02-11T21:07:35.955Z",
          project_id: 2
        }]
      }

      setup = () => {
        const utils = render(<ProjectPalette
          name={mockProps.name}
          id={mockProps.id}
          colors={mockProps.colors}
          deletePaletteFromStor={mockProps.deletePaletteFromStore}
          palettes={mockProps.palettes}
        />)
        return { utils };
      }

      const { utils } = setup();
      expect(utils).toMatchSnapshot();
    });

    test('Should run removePalette on button click', () => {
      mockProps = {
        name: 'Party Palette',
        id: 15,
        colors: ['#11232', '#11222', '#34232', '#11562', '#11852'],
        deletePaletteFromStore: jest.fn(),
        palettes: [{
          colors: ["#c005d4", "#57a3d0", "#88b8b1", "#7bac74", "#e6488b"],
          id: 15,
          name: "Candy",
          created_at: "2020-02-11T21:07:35.955Z",
          updated_at: "2020-02-11T21:07:35.955Z",
          project_id: 2
        }]
      }

      setup = () => {
        const utils = render(<ProjectPalette
          name={mockProps.name}
          id={mockProps.id}
          colors={mockProps.colors}
          deletePaletteFromStore={mockProps.deletePaletteFromStore}
          palettes={mockProps.palettes}
        />)
        const btn = utils.getByTestId('delete-button')
        return {
          btn,
          ...utils
        };
      }

      const { btn, utils } = setup();
      fireEvent.click(btn);
      expect(deletePalette).toHaveBeenCalledWith(15, {
        colors: ["#c005d4", "#57a3d0", "#88b8b1", "#7bac74", "#e6488b"],
        id: 15,
        name: "Candy",
        created_at: "2020-02-11T21:07:35.955Z",
        updated_at: "2020-02-11T21:07:35.955Z",
        project_id: 2
      });
      expect(mockProps.deletePaletteFromStore).toHaveBeenCalledWith({
        colors: ["#c005d4", "#57a3d0", "#88b8b1", "#7bac74", "#e6488b"],
        id: 15,
        name: "Candy",
        created_at: "2020-02-11T21:07:35.955Z",
        updated_at: "2020-02-11T21:07:35.955Z",
        project_id: 2
      });
    });

  describe('mapStateToProps', () => {
    it('Should return an object with the palettes', () => {
      const mockState = {
        palettes: [{
          colors: ["#c005d4", "#57a3d0", "#88b8b1", "#7bac74", "#e6488b"],
          id: 21,
          name: "Candy",
          created_at: "2020-02-11T21:07:35.955Z",
          updated_at: "2020-02-11T21:07:35.955Z",
          project_id: 2
        }],
        projects: [{name: 'Project1', id: 5}, {name: 'Project2', id: 6}]
      }
      const expected = {
        palettes: [{
          colors: ["#c005d4", "#57a3d0", "#88b8b1", "#7bac74", "#e6488b"],
          id: 21,
          name: "Candy",
          created_at: "2020-02-11T21:07:35.955Z",
          updated_at: "2020-02-11T21:07:35.955Z",
          project_id: 2
        }]
      }

      const mappedProps = mapStateToProps(mockState);
      expect(mappedProps).toEqual(expected);
    });
  });

  describe('mapDispatchToProps', () => {
    it('Should call dispatch with deletePaletteFromStore action', () => {
      const palette = {
        colors: ["#c005d4", "#57a3d0", "#88b8b1", "#7bac74", "#e6488b"],
        id: 21,
        name: "Candy",
        created_at: "2020-02-11T21:07:35.955Z",
        updated_at: "2020-02-11T21:07:35.955Z",
        project_id: 2
      };
      const mockDispatch = jest.fn();
      const actionToDispatch = deletePaletteFromStore(palette);
      const mappedProps = mapDispatchToProps(mockDispatch);

      mappedProps.deletePaletteFromStore(palette);
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
    });
  });

  });
});
