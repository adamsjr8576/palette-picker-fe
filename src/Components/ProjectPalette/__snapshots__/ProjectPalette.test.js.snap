// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProjectPalette ProjectPalette Component Should match the snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <article
        class="project-colors-article"
      >
        <div>
          <h3>
            Party Palette
          </h3>
          <button
            class="trash-btn"
            data-testid="delete-button"
          >
            <img
              class="icon-bin"
              src="bin.png"
            />
          </button>
        </div>
        <div
          class="colors-div-container"
        >
          <div
            class="color-box"
          >
            <p>
              #11232
            </p>
          </div>
          <div
            class="color-box"
          >
            <p>
              #11222
            </p>
          </div>
          <div
            class="color-box"
          >
            <p>
              #34232
            </p>
          </div>
          <div
            class="color-box"
          >
            <p>
              #11562
            </p>
          </div>
          <div
            class="color-box"
          >
            <p>
              #11852
            </p>
          </div>
        </div>
      </article>
    </div>
  </body>,
  "container": <div>
    <article
      class="project-colors-article"
    >
      <div>
        <h3>
          Party Palette
        </h3>
        <button
          class="trash-btn"
          data-testid="delete-button"
        >
          <img
            class="icon-bin"
            src="bin.png"
          />
        </button>
      </div>
      <div
        class="colors-div-container"
      >
        <div
          class="color-box"
        >
          <p>
            #11232
          </p>
        </div>
        <div
          class="color-box"
        >
          <p>
            #11222
          </p>
        </div>
        <div
          class="color-box"
        >
          <p>
            #34232
          </p>
        </div>
        <div
          class="color-box"
        >
          <p>
            #11562
          </p>
        </div>
        <div
          class="color-box"
        >
          <p>
            #11852
          </p>
        </div>
      </div>
    </article>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
